fendermint/vm/interpreter/src/chain.rs

```diff
@@ -89,32 +97,72 @@ where
             CheckpointPoolItem::BottomUp(ckpt) => ChainMessage::Ipc(IpcMessage::BottomUpExec(ckpt)),
         });
 
+        // Prepare top down proposals
+        match atomically_or_err(|| finality_provider.next_proposal()).await {
+            Ok(Some(proposal)) => msgs.push(ChainMessage::Ipc(IpcMessage::TopDownProposal(
+                ParentFinalityProposal {
+                    height: proposal.height as ChainEpoch,
+                    block_hash: proposal.block_hash,
+                },
+            ))),
+            Ok(None) => {
+                tracing::debug!("parent finality proposal not ready yet");
+            }
+            Err(e) => {
+                // safe to unwrap as the type is correct
+                let e = e.downcast_ref::<fendermint_vm_topdown::Error>().unwrap();
```

Comment:
But it is possible to use a type hint to the compiler to avoid having to cast and unwrap, right? It can be done either by assigning to a variable, or by calling it like `atomically_or_err<_, fendermint_vm_topdown::Error, _, _>(...)`.

Reply:
nice nice, thx for the tip, was not aware of this

Reply:
Sorry, I mentioned it in the description of  https://github.com/aakoshh/async-stm-rs/pull/1 which I sent to you, and then there is an example of the above at https://github.com/aakoshh/async-stm-rs/pull/1/files#diff-b613865b220392b32225c6578f96bfe2418a6d4e8d8c4c63ee664dceb7e60719R102